### PR TITLE
Speed up Position Encoding

### LINKED ISSUE #75: Speed up Position Encodings
## Detailed Description
Position encoding creation is quite slow, taking about 4 seconds for a batch of 32. To speed up our training we need to get this down.

## Context

A recent bit of profiling shows that the position encoding and all the concatenations in it take up about 4 seconds of the creation of each batch. 

```
Profile stats for: get_train_batch
         199664 function calls (194382 primitive calls) in 11.075 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     23/1    0.000    0.000   11.075   11.075 {built-in method builtins.next}
        1    0.000    0.000   11.075   11.075 supporters.py:658(prefetch_iterator)
        2    0.000    0.000   11.075    5.538 supporters.py:582(__next__)
        3    0.000    0.000   11.075    3.692 apply_func.py:69(apply_to_collection)
        2    0.000    0.000   11.075    5.537 supporters.py:591(request_next_batch)
        2    0.000    0.000   11.075    5.537 supporters.py:603(next_fn)
        2    0.000    0.000   11.075    5.537 dataloader.py:517(__next__)
        2    0.000    0.000   11.074    5.537 dataloader.py:559(_next_data)
        2    0.454    0.227   11.074    5.537 fetch.py:47(fetch)
        2    0.001    0.000   10.619    5.310 datasets.py:250(__getitem__)
        2    0.001    0.001   10.239    5.119 datasets.py:134(__getitem__)
        2    0.061    0.031    8.311    4.156 position_encoding.py:62(generate_position_encodings_for_batch)
       12    1.197    0.100    8.244    0.687 position_encoding.py:193(encode_absolute_position)
      416    0.005    0.000    4.530    0.011 einops.py:327(reduce)
      416    0.006    0.000    4.521    0.011 einops.py:202(apply)
       32    0.000    0.000    4.501    0.141 einops.py:455(repeat)
       32    0.000    0.000    4.494    0.140 _backends.py:98(add_axes)
       32    0.000    0.000    4.493    0.140 _backends.py:336(tile)
       32    4.493    0.140    4.493    0.140 {method 'repeat' of 'torch._C._TensorBase' objects}
      826    2.113    0.003    2.113    0.003 {built-in method cat}
        2    0.001    0.001    1.623    0.812 batch.py:144(load_netcdf)
       10    0.000    0.000    1.437    0.144 position_encoding.py:251(combine_space_and_time_features)
      126    1.390    0.011    1.390    0.011 {method 'acquire' of '_thread.lock' objects}
        2    0.000    0.000    1.380    0.690 _base.py:635(__exit__)
        2    0.000    0.000    1.380    0.690 thread.py:210(shutdown)
       14    0.000    0.000    1.380    0.099 threading.py:1021(join)
       14    0.000    0.000    1.379    0.099 threading.py:1059(_wait_for_tstate_lock)
       12    0.020    0.002    1.207    0.101 position_encoding.py:279(normalize_geospatial_coordinates)
      404    0.192    0.000    0.956    0.002 position_encoding.py:399(fourier_encode)
        6    0.000    0.000    0.241    0.040 datasets.py:340(add_encodings)
       14    0.000    0.000    0.207    0.015 xr_utils.py:84(validate)
        2    0.000    0.000    0.187    0.093 batch.py:157(normalize)
      404    0.187    0.000    0.187    0.000 {method 'sin' of 'torch._C._TensorBase' objects}
      396    0.182    0.000    0.182    0.000 {built-in method stack}
        4    0.000    0.000    0.162    0.041 satellite_model.py:18(model_validation)
```

## Possible Implementation
It seems like reducing the number of concatentations is the key, as generating the Fourier features is quite fast. But concatentation takes 2 of the 4 seconds it takes to create the encodings. 

The easiest way would be to just run the spatial and feature encodings over all the batch at once, not one at a time like it currently does.

An issue with that is finding some suitable replacements, for the spatial encoding, for example, `torch.meshgrid` only takes 1D tensors, and so can't be done over the whole batch at once, I think? 

For the datetime features, doing it over the whole batch should also speed it up a bit, just need to extract the hour, minute, and day of year of a set of timestamps all at once.

### PR DESCRIPTION
**Pull request by jacobbieker**
# Pull Request

## Description

Speed up the position encoding creation

Fixes #75 

## How Has This Been Tested?

Banchmarking and unit tests

- [ ] No
- [x] Yes

## Checklist:

- [x] My code follows [OCF's coding style guidelines](https://github.com/openclimatefix/nowcasting/blob/main/coding_style.md)
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by jacobbieker** [comment]
Okay, I've gotten about a 15% speed up with reducing the number of concatenations and stacks. Doing some other stuff with reducing other concatenations, etc. didn't result in any speedups, and actually slowed things down.

File: nowcasting_dataloader/utils/position_encoding.py [hunk 1]
```diff
@@ -308,18 +308,19 @@ def normalize_geospatial_coordinates(
...
     for idx in range(len(geospatial_coordinates[0])):
         x = torch.from_numpy(geospatial_coordinates[0][idx])
         y = torch.from_numpy(geospatial_coordinates[1][idx])
         grid = torch.meshgrid(x, y)
         pos = torch.stack(grid, dim=-1)
         encoded_position = fourier_encode(pos, **kwargs)
         encoded_position = einops.rearrange(encoded_position, "... n d -> ... (n d)")
-        to_concat.append(encoded_position)
+        if solid_tensor is None:
+            solid_tensor = torch.zeros(len(geospatial_coordinates[0]), *encoded_position.size())
```
Comment (JackKelly) [inline nowcasting_dataloader/utils/position_encoding.py:320]: does `torch.zeros()` use the same data type as `encoded_position`?  (to avoid having to cast between data types in line 321).

**Review by JackKelly — APPROVED** [review]
LGTM!  Well done getting a speed up!

File: nowcasting_dataloader/utils/position_encoding.py [hunk 2]
```diff
@@ -308,18 +308,19 @@ def normalize_geospatial_coordinates(
     geospatial_coordinates[1] = geospatial_coordinates[1] * 2 - 1
     # Now create a grid of the coordinates
     # Have to do it for each individual example in the batch, and zip together x and y for it
-    to_concat = []
+    solid_tensor = None
     for idx in range(len(geospatial_coordinates[0])):
         x = torch.from_numpy(geospatial_coordinates[0][idx])
```
Comment (JackKelly) [inline nowcasting_dataloader/utils/position_encoding.py:313]: Are `x` and `y` in their "minimal data types"?  (e.g. int16)?  Using the smallest data types possible might help speed things up?

Comment (JackKelly) [inline nowcasting_dataloader/utils/position_encoding.py:313]: Are `x` and `y` in their "minimal data types"?  (e.g. int16)?  Using the smallest data types possible might help speed things up?

**Review by JackKelly — APPROVED** [review]

File: nowcasting_dataloader/utils/position_encoding.py (hunk 2 above)
Comment (jacobbieker) [reply nowcasting_dataloader/utils/position_encoding.py:313]: Yeah, they are float32s I think, which they need to be as they are normalized between -1 and 1 for the grid points

File: nowcasting_dataloader/utils/position_encoding.py (hunk 1 above)
Comment (jacobbieker) [reply nowcasting_dataloader/utils/position_encoding.py:320]: Good catch! Added making it in the same dtype as `x`
